File: Binary Search/firstandlast.py
li=[1,2,3,4,5,6,7,7,8,8,8,8]

# def index(li,t):
#     ans=[]
#     low= 0
#     high = len(li)-1
#     mid = low-int((low-high)/2)
#     print(f"high {high}")
#     while low <= high:
#         if li[mid] == t:
            
#             ans.append(mid)
#             while mid <= high and li[mid]==t:
                
                
#                 mid+=1
#             ans.append(mid-1)
#             print(ans)
#             return ans[0],ans[-1]

             
                    
        
#         elif li[mid]>t:
           
#             high = mid -1
#             mid = low-int((low-high)/2)
#         else:
            
#             low = mid +1
#             mid = low-int((low-high)/2)
#     return ans
# print(index(li,8))


# better implementation

def get_first(li,t):
    low= 0
    high = len(li)-1
    mid = low-int((low-high)/2)
    ans =-1
    while low <= high:
        if li[mid] == t:
            ans=mid 
            high =mid -1
            

                     
        
        elif li[mid]>t:
           
            high = mid -1
           
        else:
            
            low = mid +1
        mid = low-int((low-high)/2)
    return ans

def get_last(li,t):
    low= 0
    high = len(li)-1
    mid = low-int((low-high)/2)
    ans =-1
    while low <= high:
        if li[mid] == t:
            ans=mid 
            low=mid +1
            

                     
        
        elif li[mid]>t:
           
            high = mid -1
           
        else:
            
            low = mid +1
        mid = low-int((low-high)/2)
    return ans
    

if __name__=='__main__':
    target = 8
    ans=[]
    ans.append(get_first(li,8))
    ans.append(get_last(li,8))
    print(ans[0],ans[1])
  

File: Binary Search/firstandlast.java
public class firstandlast {
    public static void main (String [] a){
        int target;
        target = 8;
        int [] li ={1,2,3,4,4,4,4,5,6,7,7,8,8,8,8,8,8};
        System.out.println(1/2);
        int first,last;
        first = getFirst(li,target);
        last = getLast(li, target);
        System.out.println("Occured at :"+first+ " and last at :"+last);
        System.out.println("Number of occurance is "+(last-first+1));

    }
    static int getFirst(int [] li,int t)
    {
        System.out.println("@1st");
        int l,h,m,ans;
        l=0;
        ans=-1;
        h = li.length -1;
        m = l-(l-h)/2;
        while (l<=h){
            if(t==li[m]){
                    ans = m;
                    h= m-1;
            }
            else if (li[m]<t){
                l= m+1;
            }
            else{
                h = m -1;
            }
            m = l-(l-h)/2;
        }
        return ans;
    }
    static int getLast(int [] li,int t)
    {
        System.out.println("@nds");

        int l,h,m,ans;
        l=0;
        ans=-1;
        h = li.length -1;
        m = l-(l-h)/2;
        while (l<=h){
            if(t==li[m]){
                    ans = m;
                    l= m+1;
            }
            else if (li[m]<t){
                l= m+1;
            }
            else{
                h = m -1;
            }
            m = l-(l-h)/2;
        }
        return ans;
    }

    
}
